test/sslapitest.c

```diff
@@ -5878,6 +5878,70 @@ static int test_serverinfo(int tst)
     return testresult;
 }
 
+#if !defined(OPENSSL_NO_TLS1_2) && !defined(OPENSSL_NO_DEPRECATED_3_0)
+static const unsigned char serverinfo_custom_v1[] = {
+    0x00, (char)TLSEXT_TYPE_signed_certificate_timestamp,
+    0x00, 0x03,
+    0x04, 0x05, 0x06
+};
+
+static int serverinfo_custom_v1_parse_cb(SSL *s, unsigned int ext_type,
+                                          unsigned int context,
+                                          const unsigned char *in,
+                                          size_t inlen, X509 *x,
+                                          size_t chainidx, int *al,
+                                          void *parse_arg)
+{
+    const size_t len = sizeof(serverinfo_custom_v1);
+    const unsigned char *si = &serverinfo_custom_v1[len - 3];
+    int *p_cb_result = (int*)parse_arg;
+    *p_cb_result = TEST_mem_eq(in, inlen, si, 3);
+    return 1;
+}
+
+static int test_serverinfo_custom_v1(void)
+{
+    SSL_CTX *sctx = NULL, *cctx = NULL;
+    SSL *clientssl = NULL, *serverssl = NULL;
+    int testresult = 0;
+    int cb_result = 0;
+
+    if (!TEST_true(create_ssl_ctx_pair(libctx, TLSv1_2_server_method(),
+                                       TLS_client_method(), TLS1_2_VERSION, 0,
+                                       &sctx, &cctx, cert, privkey))
+        || !TEST_true(SSL_CTX_use_serverinfo(sctx,
```

Comment:
Please add 2 more variants of this test (i.e. pass a test index into `test_serverinfo_custom_v1`, and use `ADD_ALL_TESTS` below instead of `ADD_TEST`).

We should have one variant which calls `SSL_CTX_use_serverinfo_ex` instead of `SSL_CTX_use_serverinfo` using a version of `SSL_SERVERINFOV1`, and another one which does the same thing but passes a v2 serverinfo block (`SSL_SERVERINFOV2`).

This test could then actually replace `test_serverinfo`.

My guess is that when you make that above change we will see a test failure when using `SSL_CTX_use_serverinfo_ex` and passing `SSL_SERVERINFOV1`.